devlib/target.py

```diff
@@ -2892,6 +2895,99 @@ def _resolve_paths(self):
+                 cpu_types='-cpu cortex-a72',
+                 initrd_image=str(),
+                 mem_size=512,
+                 num_cores=2,
+                 num_threads=2,
+                 cmdline='console=ttyAMA0',
+                 enable_kvm=True,
+                 boot_timeout=60,
+                 ):
+        super().__init__(connection_settings=connection_settings,
+                         platform=platform,
+                         working_directory=working_directory,
+                         executables_directory=executables_directory,
+                         connect=False,
+                         modules=modules,
+                         load_default_modules=load_default_modules,
+                         shell_prompt=shell_prompt,
+                         conn_cls=SshConnection,
+                         max_async=max_async)
+
+        qemu_executable = f'qemu-system-{arch}'
+        qemu_path = shutil.which(qemu_executable)
+        if qemu_path is None:
+            raise FileNotFoundError(f'Cannot find {qemu_executable} executable!')
+
+        if not os.path.exists(kernel_image):
+            raise FileNotFoundError(f'{kernel_image} does not exist!')
+
+        qemu_cmd = f'sudo {qemu_path} -kernel {kernel_image} -append "{cmdline}" ' \
+                   f'-smp cores={num_cores},threads={num_threads} -m {mem_size} ' \
+                   f'-netdev user,id=net0,hostfwd=tcp::{connection_settings["port"]}-:22 ' \
+                   '-device virtio-net-pci,netdev=net0 --nographic'
+
+        if initrd_image:
+            if not os.path.exists(initrd_image):
+                raise FileNotFoundError(f'{initrd_image} does not exist!')
+            qemu_cmd = f'{qemu_cmd} -initrd {initrd_image}'
+
+        if arch.startswith('x86'):
+            qemu_cmd = f'{qemu_cmd} {"--enable-kvm" if enable_kvm else ""}'
+        else:
+            qemu_cmd = f'{qemu_cmd} -machine virt {cpu_types}'
+
+        self.logger.debug('qemu_cmd: %s', qemu_cmd)
+
+        try:
+            self.qemu_process = get_subprocess(qemu_cmd, shell=True)
+        except Exception as ex:
+            raise HostError(f'Error while running "{qemu_cmd}": {ex}') from ex
+
+        self.wait_boot_complete(timeout=boot_timeout)
+
+        if connect:
+            super().connect(check_boot_completed=False)
+
+    def wait_boot_complete(self, timeout=10):
+        start_time = time.time()
+        boot_completed = False
+        poll_obj = select.poll()
+        poll_obj.register(self.qemu_process.stdout, select.POLLIN)
+
+        while not boot_completed and timeout >= (time.time() - start_time):
+            poll_result = poll_obj.poll(0)
+            if poll_result:
+                line = self.qemu_process.stdout.readline().rstrip(b'\r\n')
+                line = line.decode(sys.stdout.encoding or 'utf-8', "replace") if line else ''
+                self.logger.debug(line)
+                if 'Welcome to Buildroot' in line:
```

Comment:
I think it indicates SSH server is ready:
```Bash
ssh-keygen: generating new host keys: RSA ECDSA ED25519 
Starting sshd: OK

Welcome to Buildroot
```
But it's mainly for waiting boot completion (so the method name is `wait_boot_complete()`).

I did not like hard-coding the banner, too, but port scanning looks overkill. Maybe it should be parametric. If a user customizes the default image or has an existing image with a different banner, they can specify that.

Reply:
I don't think it's overkill, it's not even more code:
https://stackoverflow.com/questions/19196105/how-to-check-if-a-network-port-is-open
The banner will change if we try to use e.g. Debian, and it can probably include variable things like the date.